postq.go

```diff
@@ -0,0 +1,176 @@
+package postq
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/commons/utils"
+)
+
+const (
+	// eventMaxAttempts is the maximum number of attempts to process an event in `event_queue`
+	EventMaxAttempts = 3
+
+	// eventQueueUpdateChannel is the channel on which new events on the `event_queue` table
+	// are notified.
+	EventQueueUpdateChannel = "event_queue_updates"
+)
+
+type (
+	// AsyncEventHandlerFunc processes multiple events and returns the failed ones
+	AsyncEventHandlerFunc func(Context, []Event) []Event
+
+	// SyncEventHandlerFunc processes a single event and ONLY makes db changes.
+	SyncEventHandlerFunc func(Context, Event) error
+)
+
+// fetchEvents fetches given watch events from the `event_queue` table.
+func fetchEvents(ctx Context, watchEvents []string, batchSize int) ([]Event, error) {
+	const selectEventsQuery = `
+			DELETE FROM event_queue
+			WHERE id IN (
+				SELECT id FROM event_queue
+				WHERE 
+					attempts <= @maxAttempts AND
+					name IN @events AND
+					(last_attempt IS NULL OR last_attempt <= NOW() - INTERVAL '1 SECOND' * @baseDelay * POWER(attempts, @exponential))
+				ORDER BY priority DESC, created_at ASC
+				FOR UPDATE SKIP LOCKED
+				LIMIT @batchSize
+			)
+			RETURNING *
+		`
+
+	var events []Event
+	vals := map[string]any{
+		"maxAttempts": EventMaxAttempts,
```

Comment:
I think maxAttempts, delay, exponential should all be parameters on EventConsumers with defaults